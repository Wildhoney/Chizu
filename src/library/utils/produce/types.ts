import { Model, State } from "../../types";
import { validatable } from "./utils";

export class Models<M extends Model> {
  constructor(
    public stateless: M,
    public stateful: M = stateless,
  ) {}

  get validatable(): Validatable<M> {
    return validatable(this.stateful);
  }
}

export type Validate = {
  pending(): boolean;
  is(operation: State): boolean;
  draft<T>(): T;
};

export type Validatable<M> = {
  [K in keyof M]: M[K] extends object
    ? M[K] extends Array<infer U>
      ? Array<Validatable<U> & Validate> & Validate
      : Validatable<M[K]> & Validate
    : M[K] & Validate;
} & Validate;
