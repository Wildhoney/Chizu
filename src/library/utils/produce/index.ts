import { ModuleDefinition } from "../../types/index.ts";
import Maybe from "../maybe/index.ts";
import clone from "rfdc";

export type MaybeObject = { __kind: "Maybe" };

export type Producible<T> = {
  [K in keyof T]: T[K] extends Maybe<infer U>
    ? U extends object
      ? U & MaybeObject
      : T[K]
    : T[K];
};

export default function produce<M extends ModuleDefinition["Model"]>(
  model: M,
): [Producible<M>, Producible<M>] {
  function next(model: Producible<M>): Producible<M> {
    return new Proxy(model, {
      get(target, prop) {
        const value = Reflect.get(target, prop) as Producible<M>;
        if (typeof prop === "symbol") return value;
        if (value === undefined) Reflect.set(target, prop, {});
        return next(Reflect.get(target, prop) as Producible<M>);
      },

      set(target, prop, value) {
        Reflect.set(target, prop, value);
        return true;
      },
    });
  }

  // const cloned = clone({ proto: true })(model) as Producible<M>;
  // console.log(model, cloned);
  const proxy = next(model);
  return [model, proxy];
}
