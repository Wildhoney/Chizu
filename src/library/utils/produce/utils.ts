import { Process, State } from "../../types/index.ts";
import { Immer } from "immer";

export const config = {
  immer: new Immer(),
  states: "___states",
};

config.immer.setAutoFreeze(false);

export class Stateful<M> {
  process: null | Process;

  constructor(
    public value: M,
    public operations: State[],
    public field: null | number | string = null,
  ) {
    this.process = null;
  }

  public attach(process: Process): Stateful<M> {
    this.process = process;
    return this;
  }
}

export function state<M>(value: M, operations: State[] = []): M {
  return new Stateful(value, operations) as M;
}
