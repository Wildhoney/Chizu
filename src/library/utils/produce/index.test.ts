import produce from "./index.ts";
import { describe, expect, it } from "@jest/globals";

describe("produce", () => {
  it("should create a proxy object", () => {
    type Model = {
      name: { first: string; last: string };
      spouse: { name: { first: string } };
      children: { names: string[] };
    };

    const [model, proxy] = produce<Model>({
      name: { first: "", last: "Timberlake" },
      spouse: { name: { first: "" } },
      children: { names: [] },
    });

    proxy.name.first = "Adam";
    proxy.spouse.name.first = "Maria";
    proxy.children.names = ["Imogen", "Phoebe"];

    expect(model).toMatchSnapshot();
  });
});
