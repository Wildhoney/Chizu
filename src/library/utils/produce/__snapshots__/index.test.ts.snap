// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`produce should create a proxy object 1`] = `
{
  "children": {
    "names": [
      "Imogen",
      "Phoebe",
    ],
  },
  "name": {
    "first": "Adam",
    "last": "Timberlake",
  },
  "spouse": {
    "name": {
      "first": "Maria",
    },
  },
}
`;
